py/src/nupic/algorithms/anomaly_likelihood.py

```diff
@@ -0,0 +1,680 @@
+
+  # Update the estimator
+  newParams = {
+    "distribution": params["distribution"],
+    "movingAverage": {
+      "historicalValues": historicalValues,
+      "total": total,
+      "windowSize": windowSize,
+    },
+    "historicalLikelihoods": historicalLikelihoods,
+  }
+
+  assert len(newParams["historicalLikelihoods"]) <= windowSize
+
+  if verbosity > 3:
+    print(("Number of likelihoods:", len(likelihoods)))
+    print(("First 20 likelihoods:", likelihoods[0:min(20, len(likelihoods))]))
+    print("Leaving updateAnomalyLikelihoods.")
+
+  return (likelihoods, aggRecordList, newParams)
+
+
+
+def _filterLikelihoods(likelihoods,
+                       redThreshold=0.99999, yellowThreshold=0.999):
+  """
+  Filter the list of raw (pre-filtered) likelihoods so that we only preserve
+  sharp increases in likelihood. 'likelihoods' can be a numpy array of floats or
+  a list of floats.
+
+  :returns: A new list of floats likelihoods containing the filtered values.
+  """
+  redThreshold    = 1.0 - redThreshold
+  yellowThreshold = 1.0 - yellowThreshold
+
+  # The first value is untouched
+  filteredLikelihoods = [likelihoods[0]]
+
+  for i, v in enumerate(likelihoods[1:]):
+
+    if v <= redThreshold:
+      # Value is in the redzone
+
+      if likelihoods[i] > redThreshold:
+        # Previous value is not in redzone, so leave as-is
+        filteredLikelihoods.append(v)
+      else:
+        filteredLikelihoods.append(yellowThreshold)
+
+    else:
+      # Value is below the redzone, so leave as-is
+      filteredLikelihoods.append(v)
+
+  return filteredLikelihoods
+
+
+
+def _anomalyScoreMovingAverage(anomalyScores,
+                               windowSize=10,
+                               verbosity=0,
+                              ):
+  """
+  Given a list of anomaly scores return a list of averaged records.
+  anomalyScores is assumed to be a list of records of the form:
+                [datetime.datetime(2013, 8, 10, 23, 0), 6.0, 1.0]
+
+  Each record in the returned list list contains:
+      [datetime, value, averagedScore]
+
+  *Note:* we only average the anomaly score.
+  """
+
+  historicalValues = []
+  total = 0.0
+  averagedRecordList = []    # Aggregated records
+  for record in anomalyScores:
+
+    # Skip (but log) records without correct number of entries
+    if not isinstance(record, (list, tuple)) or len(record) != 3:
+      if verbosity >= 1:
+        print(("Malformed record:", record))
+      continue
+
+    avg, historicalValues, total = (
+      MovingAverage.compute(historicalValues, total, record[2], windowSize)
+      )
+
+    averagedRecordList.append( [record[0], record[1], avg] )
+
+    if verbosity > 2:
+      print(("Aggregating input record:", record))
+      print(("Result:", [record[0], record[1], avg]))
+
+  return averagedRecordList, historicalValues, total
+
+
+
+def estimateNormal(sampleData, performLowerBoundCheck=True):
```

Comment:
I'm pretty sure that numpy/scipy have library functions to do this, which return objects with all sorts of conveniences attached.

Reply:
First of all, I want to say I really appreciate you do deep code review where you understand the code and question the implementation! :+1: 

As for many of these comments, I agree, but I don't want to act on them in this PR which just brings 1:1 code from numenta's, which can be considered for sort of a standard. 

> I'm pretty sure that numpy/scipy have library functions to do this, which return objects with all sorts of conveniences attached.

as for the justified changes. Yes, the code could (should?) be improved. The result of your review should be fixed in nupic.py (and then brought back here?) 

Q: do we want to deviate from Numenta's python implementation? (we said (only API-wise?) we'll keep strong compatibility. 

Q: the nupic repo is definitely more alive than nupic.core, should we offer fix there? @rhyolight will you accept it? 

Q: actually, do we want to (long term) have PY code here, shouldn't we just focus on API compat with numenta/nupic or community/nupic.py? (I think it's fine to merge all py code in this repo)

Reply:
see https://github.com/numenta/nupic/issues/3876